h2/src/main/org/h2/mvstore/Page.java

```diff
@@ -387,30 +382,9 @@ protected void dump(StringBuilder buff) {
      * @return the value or null
      */
     int binarySearch(K key) {
-        int low = 0, high = getKeyCount() - 1;
-        // the cached index minus one, so that
-        // for the first time (when cachedCompare is 0),
-        // the default value is used
-        int x = cachedCompare - 1;
-        if (x < 0 || x > high) {
-            x = high >>> 1;
-        }
-        K[] k = keys;
-        DataType<K> keyType = map.getKeyType();
-        while (low <= high) {
-            int compare = keyType.compare(key, k[x]);
-            if (compare > 0) {
-                low = x + 1;
-            } else if (compare < 0) {
-                high = x - 1;
-            } else {
-                cachedCompare = x + 1;
-                return x;
-            }
-            x = (low + high) >>> 1;
-        }
-        cachedCompare = low;
-        return -(low + 1);
+        int res = map.getKeyType().binarySearch(key, keys, getKeyCount(), cachedCompare);
+        cachedCompare = (res < 0 ? -res : res) - 1;
```

Comment:
Both old an new logic expect `cachedCompare` to be within the range `[1, size]` for cached path.

The previous logic generated the `cachedCompare` like
```Java
cachedCompare = res < 0 ? ~res : res + 1;
```
Here the
```Java
cachedCompare = res < 0 ? ~res : res - 1;
```
is used.

When entry isn't found `~res` produces values within the range `[0, size]`, this range is wider than the expected range, so the fallback to default may be unexpectedly used by both old and new code.

When an entry is found, old `res + 1` produced values within the range `[1, size]`, it was correct. This range is expected and `cachedCompare - 1` is actually used, so `res + 1 - 1 = res` (previous entry) was used as the starting point.

When an entry is found, new `res - 1` produces the values within the range `[-1, size - 2]`, it looks like off-by-two error.

So one of two paths was buggy and now both paths are buggy (suboptimal).

Maybe I'm missing something, please re-check.

(`~something == -something - 1`)

Reply:
When entry isn't found the bitwise negation of insertion point is returned. The insertion point has larger range than number of keys.

If the insertion point is 0, we should start the search from the 0 (`cachedCompare = 1`). If the insertion point is `size`, we should start the search from the `size - 1` (`cachedCompare = size`).

If the insertion point is 1, we should start the search from the 0 too, I guess, as the old and the new code. So we need an additional branch for the insertion point 0 (`~0 == -1`).

```Java
cachedCompare = res < 0 ? res == -1 ? 1 : ~res : res + 1;
//                        ^^^^^^^^^^^^^              ^^^
//                        special case         off-by-two fix
//                     (for insert before)
```